sla-miss-report/sla_miss_report.py

```diff
@@ -0,0 +1,902 @@
+import json
+from dataclasses import dataclass
+from datetime import date, datetime, timedelta
+from distutils.command.sdist import sdist
+from pprint import pprint
+
+import airflow
+import numpy as np
+import pandas as pd
+from airflow import settings, utils
+from airflow.models import DAG, DagRun, SlaMiss, TaskInstance
+from airflow.models.serialized_dag import SerializedDagModel
+from airflow.operators.email import EmailOperator
+from airflow.operators.python_operator import PythonOperator
+from airflow.utils.session import NEW_SESSION, create_session, provide_session
+from airflow.version import version
+from sqlalchemy import (
+    Boolean,
+    Column,
+    ForeignKey,
+    Index,
+    Integer,
+    String,
+    Text,
+    create_engine,
+    func,
+    not_,
+    or_,
+)
+from sqlalchemy.orm import backref, joinedload, relationship
+from sqlalchemy.orm.query import Query
+from sqlalchemy.orm.session import Session
+from yaml import serialize
+
+session = settings.Session()
+
+slamiss = session.query(
+    SlaMiss.task_id, SlaMiss.dag_id, SlaMiss.execution_date, SlaMiss.timestamp
+).all()
+slamiss_df = pd.DataFrame(slamiss)
+slamiss_df["run_date"] = pd.to_datetime(slamiss_df["execution_date"]).dt.date
+slamiss_df["run_date_hour"] = pd.to_datetime(slamiss_df["execution_date"]).dt.hour
+
+taskinstance = session.query(
+    TaskInstance.task_id,
+    TaskInstance.dag_id,
+    TaskInstance.run_id,
+    TaskInstance.state,
+    TaskInstance.start_date,
+    TaskInstance.end_date,
+    TaskInstance.duration,
+    TaskInstance.operator,
+    TaskInstance.queued_dttm,
+).all()
+taskinstance_df = pd.DataFrame(taskinstance)
+taskinstance_df["run_date"] = pd.to_datetime(taskinstance_df["start_date"]).dt.date
+taskinstance_df["run_date_hour"] = pd.to_datetime(taskinstance_df["start_date"]).dt.hour
+taskinstance_df["task_queue_time"] = (
+    taskinstance_df["start_date"] - taskinstance_df["queued_dttm"]
+).dt.total_seconds()
+taskinstance_df = taskinstance_df[taskinstance_df["task_queue_time"] > 0]
+
+dagrun = session.query(DagRun.dag_id, DagRun.run_id, DagRun.data_interval_end).all()
+dagrun_df = pd.DataFrame(dagrun)
+dagrun_df.rename(columns={"data_interval_end": "actual_start_time"}, inplace=True)
+# print(dagrun_df[:10])
+# dagrun.rename(columns={'gdp':'log(gdp)'}, inplace=True)
+# taskinstance_df.info()
+
+serializeddag = session.query(SerializedDagModel._data).all()
+serializeddag_dfjson = pd.DataFrame(serializeddag)
+df2 = serializeddag_dfjson["_data"].apply(json.dumps)
+df3 = df2.apply(json.loads)
+df4 = pd.DataFrame(df3.values.tolist())["dag"]
```

Comment:
try to see if you can consolidate all these transformations (line 71-74) into one line